Vdimitrakis/add price in posts endpoint

therocket-gr opened the pull request:
We need to expose the WooCommerce price in the blaze dashboard list (in Advertising - or in the new  Marketing page)

This change will check if the  `wc_get_product ` method is available and will call for each of the items/posts, it will retrieve the price of each post (if a product) and will inject it in the final results.

This will only work for self-hosted or pressable sites.

<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes #

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
*

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

You need to have Woo installed in your testing site and to have (at least) a product created. 
Navigate to the advertising page and check your network inspect tab. Filter requests by /blaze/posts
In the request result, you should see the price data when the post is a product.

![Screenshot 2024-01-17 at 17 01 50](https://github.com/Automattic/jetpack/assets/1416426/a9e466d4-3f66-4ff7-af08-9d4ad24ecd6a)

Then click on the Promote button (on the top right corner of your page). Click continue to load all posts again and check the network for the same call.

you should see the price entry in the posts/results array

posts that are not of product type should have the property but with an empty value (empty string) 


* Go to '..'
*

therocket-gr committed:
test commit

therocket-gr committed:
test commit

therocket-gr committed:
test commit

therocket-gr committed:
Add price in /blaze/posts endpoint

therocket-gr committed:
changelog entry

therocket-gr committed:
strip_tags in price result

therocket-gr committed:
fixup-project-versions

therocket-gr committed:
changelog

sbarbosa commented on projects/packages/blaze/src/class-dashboard-rest-controller.php:
```diff
@@ -320,11 +320,24 @@ public function get_blaze_posts( $req ) {
+			return $res;
+		}
+
+		foreach ( $res['posts'] as $key => $post ) {
```
The response may be an error, so I think it would be nice to check that we got a 200 response before parsing the body. 

We have some utility methods to check if the response was an error, get the response code and the response body:
```
is_wp_error
wp_remote_retrieve_response_code
wp_remote_retrieve_body
```

Maybe some of those will help you in this code.

therocket-gr replied:
both comments are fixed and code is updated.. thanks for testing it out

sbarbosa reviewed (commented):
Hi @therocket-gr! 
I checked the response and I am getting some scaped HTML code on it. I would be nice if we can clean that up. Just to be more compatible with other clients like Mobile apps.

![Screenshot 2024-01-16 at 14 31 50](https://github.com/Automattic/jetpack/assets/1258162/2a6e3f8c-15b5-41e8-ae26-a52d96299cf0)

We are already cleaning up all HTML tags, maybe we can also run `html_entity_decode` to decode the rest of the code. That function should convert `&#36;&nbsp;` to the correct `$ `

therocket-gr committed:
Fixes in foreach loop
- created a function to reuse the logic
- added the same functionality in get_dsp_blaze_posts

therocket-gr committed:
Added ENT_QUOTES in the html_entity_decode php function

sbarbosa reviewed (dismissed):
Thanks for the change Vassilis! 
I made a couple of changes to the dashboard to display the new price, and the changes in the endpoint are working great 

![Screenshot 2024-01-17 at 14 02 26](https://github.com/Automattic/jetpack/assets/1258162/cee54ba9-9926-4331-bcb5-aad247ad2084)

jeherve commented on projects/packages/blaze/src/class-dashboard-rest-controller.php:
```diff
@@ -639,6 +659,35 @@ public function edit_dsp_generic( $path, $req, $args = array() ) {
+		foreach ( $posts as $key => $post ) {
+			$product = wc_get_product( $post['ID'] );
+			if ( $product !== false ) {
+				$posts[ $key ]['price'] = html_entity_decode(
+					wp_strip_all_tags(
+						wc_price( $product->get_price() )
+					),
+					ENT_QUOTES
+				);
```
I take it there is no simple function to retrieve a formatted and escaped product price?

therocket-gr replied:
@jeherve  sorry for the delay in responding... we were in a division meetup.

i did some research and asked around, I changed the code (even though it is more lines of code) i think it is better to not use entity_decode and strip tags

@sbarbosa please let me know what you think as well

jeherve reviewed (commented):
This makes sense. I only have a minor question, and I believe this will need a rebase.

therocket-gr committed:
Remove the wc_price and format the price using WC functions

therocket-gr committed:
Merge branch 'trunk' into vdimitrakis/add-price-in-posts-endpoint

# Conflicts:
#	projects/packages/blaze/package.json
#	projects/packages/blaze/src/class-dashboard.php

PanosSynetos commented on projects/packages/blaze/src/class-dashboard-rest-controller.php:
```diff
@@ -639,6 +659,55 @@ public function edit_dsp_generic( $path, $req, $args = array() ) {
+			return $posts;
+		}
+
+		foreach ( $posts as $key => $post ) {
```
As a general advice, avoid using `$post` as it's "reserved" by `global $post` and can cause confusion.

PanosSynetos commented on projects/packages/blaze/src/class-dashboard-rest-controller.php:
```diff
@@ -639,6 +659,55 @@ public function edit_dsp_generic( $path, $req, $args = array() ) {
+			return $posts;
+		}
+
+		foreach ( $posts as $key => $post ) {
```
As a general advice, avoid using `$post` as it's "reserved" by `global $post` and can cause confusion.

therocket-gr replied:
ok changed

PanosSynetos commented on projects/packages/blaze/src/class-dashboard-rest-controller.php:
```diff
@@ -639,6 +659,55 @@ public function edit_dsp_generic( $path, $req, $args = array() ) {
+		}
+
+		foreach ( $posts as $key => $post ) {
+			$product = wc_get_product( $post['ID'] );
```
Maybe play defensive here, and check if `!isset( $post['ID']` and `continue`

Maybe use YODA syntax? :)

```
if ( false !== $product ) {
```
The product can be `object`,`null` or `false`

If the product is `null`, the above check would `return true` and will not go to `else`.

Maybe consider using positive logic.


```
if ( $product ) {
```

Oh, and I'd remove the empty new lines 693,695 etc

therocket-gr replied:
fixed

PanosSynetos reviewed (dismissed):
Hey @therocket-gr  - As we discussed over DMs, you followed the same approach as what `wc_price` is doing. 

I like the defensive coding, maybe you're overdoing it with checking all the functions, but better safe than sorry :)

Pre-approving this in order not to block you, my comments are minor

FYI - didn't actually tested the endpoint, I just reviewed the PHP side of getting the prices

therocket-gr committed:
Fixed code in relation to gh comments

therocket-gr committed:
Merge branch 'trunk' into vdimitrakis/add-price-in-posts-endpoint

# Conflicts:
#	projects/packages/blaze/package.json
#	projects/packages/blaze/src/class-dashboard.php

therocket-gr committed:
decode html special chars

therocket-gr committed:
Adding ENT_COMPAT

therocket-gr committed:
change to html_entity_decode

therocket-gr committed:
Added ENT_COMPAT again to be PHP 8.1 compatible

sbarbosa reviewed (dismissed):
I re-tested the changes and they are working fine 👍

![Screenshot 2024-02-15 at 10 39 17](https://github.com/Automattic/jetpack/assets/1258162/591e10ca-99bd-4f14-90b6-81953f28e7da)
![Screenshot 2024-02-15 at 10 39 24](https://github.com/Automattic/jetpack/assets/1258162/11d0d11e-9fbb-4678-b45d-4d85cd17113d)

therocket-gr committed:
Merge branch 'trunk' into vdimitrakis/add-price-in-posts-endpoint

therocket-gr commented:
@jeherve i think we are good to go!! I refactored a bit the code on how the price is fetched. I would appreciate your thoughts and if you think we are ok i will merge this one. Thanks in advance

jeherve commented on projects/packages/blaze/src/class-dashboard-rest-controller.php:
```diff
@@ -709,6 +729,54 @@ public function edit_dsp_generic( $path, $req, $args = array() ) {
+				continue;
+			}
+			$product = wc_get_product( $item['ID'] );
+			if ( $product ) {
```
You may have to be more defensive here, just like in #35566, to avoid fatals in some scenarios.

therocket-gr replied:
@jeherve check out this comment by Panos

https://github.com/Automattic/jetpack/pull/35066#discussion_r1488101621

i think we should be ok

jeherve replied:
I'm afraid that's not going to be enough. In some scenarios, Woo extensions may change what's returned there. That's something we ran into a few times in the past, and why we ended up checking for `$product instanceof WC_Product` in the PR I linked to earlier. Also see #11301 for another example.

therocket-gr replied:
Ok fixed, added the extra condition

PanosSynetos replied:
Being extra defensive doesn't hurt :)

jeherve reviewed (commented):
This tests well for me. I only have one small remark to ensure compatibility with other Woo extensions.

therocket-gr committed:
check if product is an instance of WC_Product in add_prices_in_posts

PanosSynetos reviewed (approved):
Code reviewed - being extra defensive is good!

therocket-gr commented:
Thank you both @PanosSynetos and @jeherve for reviewing and approving it...
